File: syllabus-1/day-4/MINWALK.cpp
#include <bits/stdc++.h>
using namespace std;

#define endl "\n"
#define ll long long
#define pairil pair<ll,int>

void djs(vector<vector<pair<int, int>>>& arr, int st, vector<ll>& costs) {
	vector<bool> visited(arr.size(), 0);

	priority_queue<pairil, vector<pairil>, greater<pairil>> q;
	q.push({ 0,st });
	costs[st] = 0;

	while (q.empty() == false) {
		pairil temp = q.top();
		q.pop();

		ll val = temp.first;
		int node = temp.second;

		if (visited[node] == 1) continue;

		// cout<<node + 1<<endl;

		visited[node] = 1;
		for (auto i : arr[node]) {
			if (visited[i.first] == 0) {
				if (costs[i.first] == -1) {
					costs[i.first] = i.second + val;
				}
				else {
					costs[i.first] = min(costs[i.first], (ll)i.second + val);
				}
				q.push({ costs[i.first],i.first });
			}
		}
	}
}
void solve() {
	int n, e;
	cin >> n >> e;

	int s, v, t;
	cin >> s >> v >> t;
	s--;    v--;    t--; // to make nodes 0 indexed

	vector<vector<pair<int, int>>> arr(n);
	for (int i = 0;i < e;++i) {
		int a, b, c;
		cin >> a >> b >> c;
		a--;    b--; // to make nodes 0 indexed

		arr[a].push_back({ b,c });
		arr[b].push_back({ a,c });
	}

	vector<ll> sCosts(n, -1), vCosts(n, -1), tCosts(n, -1);
	djs(arr, s, sCosts);
	djs(arr, v, vCosts);
	djs(arr, t, tCosts);

	ll res = -1;
	for (int i = 0;i < n;++i) {
		ll val = sCosts[i] + vCosts[i] + tCosts[i];
		if (res == -1 || res > val) {
			res = val;
		}
	}
	cout << res << endl;
}

int main() {
	ios_base::sync_with_stdio(false);cin.tie(NULL);cout.tie(NULL);
	int t = 1;
	cin >> t;
	while (t--) {
		solve();
	}

	return 0;
}

File: syllabus-1/day-4/MINWALK.py
import heapq


def dfs(st, arr):
    costs = [-1 for i in range(len(arr))]
    visited = [0 for i in range(len(arr))]

    q = []
    heapq.heappush(q, (0, st))
    costs[st] = 0

    while(len(q) > 0):
        temp = heapq.heappop(q)
        val = temp[0]
        node = temp[1]

        if(visited[node] == 1):
            continue

        visited[node] = 1
        for i in arr[node]:
            if(visited[i[0]] != 0):
                continue

            if(costs[i[0]] == -1):
                costs[i[0]] = i[1] + val
            else:
                costs[i[0]] = min(costs[i[0]], i[1] + val)
            heapq.heappush(q, (costs[i[0]], i[0]))
    return costs


def solve():
    (n, e) = map(int, input().split())
    (s, v, t) = map(int, input().split())
    s -= 1
    v -= 1
    t -= 1

    arr = [[] for i in range(n)]
    for i in range(e):
        (a, b, c) = map(int, input().split())
        a -= 1
        b -= 1

        arr[a].append((b, c))
        arr[b].append((a, c))

    sCosts = dfs(s, arr)
    vCosts = dfs(v, arr)
    tCosts = dfs(t, arr)

    res = -1
    for i in range(n):
        val = sCosts[i] + vCosts[i] + tCosts[i]
        if(res == -1 or res > val):
            res = val
    print(res)


# testcase
t = int(input())
while(t > 0):
    solve()
    t -= 1
